### PR TITLE
Clear Comments Feature

### LINKED ISSUE #25: method to clear the cache/comments of a version/guid
currently, to remove all the comments of a version the user must do so manually.

### PR DESCRIPTION
**Pull request by chrstinalin**
Fixes #25

**Changes**
- Manually deleting comments from context
image
- Setting to delete comments automatically on export. The first time the user exports, Assay will prompt for whether the user wants to delete the comments in the version after exporting. It will then ask the user if they would like to save their choice. On future exports, Assay will immediately delete, ignore, or ask to delete comments according to user preference. This preference can be changed again in settings.
image
image
image
- Fixed issue where exporting from context would export from the active text editor rather than whichever folder was selected.

### CODE REVIEW

**Review by dotproto — CHANGES_REQUESTED** [review]
I think it's out of scope for this PR, but the main through-line on my comments in this review is that in the future I think we should adopt a more object-oriented approach to Assay's architecture. I mentioned in a previous review that I thought we should collect some related functionality in a CommentManager class. I see a couple other areas where can start to collect functionality into classes, namely Comment and CommentCollection. It might be best if we chat more about this over a call. 

The suggested changes to test descriptions are primarily for consistency with how existing tests are written. To be honest, I'd prefer to have all tests use a terminal period, but I don't want to roll that change into this PR. If you'd like to change the pattern to use terminal periods, feel free to reject my suggestions here and create a new PR that introduce terminal periods in the other tests.

File: src/commands/exportComments.ts [hunk 1]
```diff
@@ -62,5 +39,31 @@ export async function exportVersionComments(uri: vscode.Uri) {
...
+) {
+  const deleteCachedComments = await getDeleteCommentsPreference();
+
+  const document = await vscode.workspace.openTextDocument({
+    content: compiledComments,
+    language: "text",
+  });
+
+  const edit = new vscode.WorkspaceEdit();
+  vscode.workspace.applyEdit(edit);
```
[the remark this answers is not in the document]
Comment (chrstinalin) [reply src/commands/exportComments.ts:57]: I'm not sure, actually. This was pre-existing code I glossed over originally. It doesn't really make sense to me either (and I cant see it affecting the function), so I just removed it.

File: src/utils/commentManager.ts [hunk 2]
```diff
@@ -218,11 +271,20 @@ export class CommentManager {
...
+    const comments = await this.fetchCommentsFromCache();
+    return this.iterateByComment(comments);
+  }
+
   /**
    * Iterates through each comment in cache.
    * @param comments The raw cache object.
    */
-  private *iterateComments(comments: CommentsCache) {
+  private *iterateByComment(comments: CommentsCache) {
```
[the remark this answers is not in the document]
Comment (chrstinalin) [reply src/utils/commentManager.ts:288]: I think refactoring Assay to be more object-oriented may be an issue in of itself. I'm finding the current folder structure less and less suitable, and I feel fully committing to OOP should come with refactoring the folders as well.

File: src/commands/exportComments.ts [hunk 3]
```diff
@@ -62,5 +48,29 @@ export async function exportVersionComments(uri: vscode.Uri) {
...
+  });
+
+  vscode.window.showTextDocument(document, vscode.ViewColumn.Beside);
+
+  if (compiledComments) {
+    vscode.env.clipboard.writeText(compiledComments);
+    vscode.window.showInformationMessage("Comments copied to clipboard.");
+  }
+
+  if (deleteCachedComments) {
```
Comment (dotproto) [inline src/commands/exportComments.ts:72]: Nit: If the variable isn't going to be used until later, it's best to move it's declaration closer to it's use location. This helps with readability as the reader doesn't have to hold as much state in mind while reading through the code. It also has the (generally very minor) benefit that an early return will not perform unnecessary work.
```suggestion
  const document = await vscode.workspace.openTextDocument({
    content: compiledComments,
    language: "text",
  });

  vscode.window.showTextDocument(document, vscode.ViewColumn.Beside);

  if (compiledComments) {
    vscode.env.clipboard.writeText(compiledComments);
    vscode.window.showInformationMessage("Comments copied to clipboard.");
  }

  const deleteCachedComments = await getDeleteCommentsPreference();
  if (deleteCachedComments) {
```

**Review by dotproto — APPROVED** [review]
LGTM. I shared a couple of thoughts, but none are blocking.

In the future I'd prefer to keep changes like the `.test.ts` terminal period changes in separate commits. Some notable disadvantages of rolling those changes into this PR rae that it increases the amount of code that needs to be reviewed in the PR, bloats the number of changes and scope of final merge commit, and makes it more difficult to understand what's actually changing in the commit when digging through version history or a `git blame`.

File: src/utils/getDeleteCommentsPreference.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+  return selectedPreference === "Yes";
+}
+
+async function setdeleteCommentsPreference(
+  config: vscode.WorkspaceConfiguration,
+  selectedPreference: string
+) {
+  const input = await vscode.window.showQuickPick(
+    ["Save my Preference", "Ask Every Time"],
```
[the remark this answers is not in the document]
Comment (chrstinalin) [reply src/utils/getDeleteCommentsPreference.ts:42]: With code-manager, there's currently a hardcoded link in addons-server that directly links the English version of it. Though, now that I'm looking at it, seems more a [limitation](https://github.com/mozilla/addons-server/blob/8f08af88323bbb871b51dda42be39d4c1db0021f/src/olympia/reviewers/templatetags/code_manager.py#L10) than a requirement.

This feels like good practice to implement regardless, so I made the issue anyway.

**Pull request merged**